{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from sys import stderr, stdout\n",
    "import tempfile\n",
    "import subprocess\n",
    "from warnings import catch_warnings\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO, SeqRecord, Seq, SearchIO, AlignIO, Phylo\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "import Bio.Entrez\n",
    "from Bio.Phylo.TreeConstruction import DistanceCalculator,DistanceTreeConstructor\n",
    "from pandas.core.frame import DataFrame\n",
    "import logging\n",
    "import time\n",
    "import traceback"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to open fasta file of imput\n",
    "def open_fasta(filename) -> SeqRecord:\n",
    "    with open(filename) as handle:\n",
    "        sequence_record = SeqIO.read(handle, 'fasta')\n",
    "    logging.info('Opened sequence {}'.format(sequence_record.id))\n",
    "    return sequence_record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to run BLAST with taxid list\n",
    "def blastp_with_list(sequence, list_taxid = [], query_size = 200):\n",
    "    result_handler, result_storer = None, None\n",
    "    #If list is empty run query without specific taxid\n",
    "    if len(list_taxid) <1:\n",
    "        result_handler = NCBIWWW.qblast('blastp', 'nr', sequence, hitlist_size=query_size) \n",
    "        result_storer = result_handler.read()\n",
    "    #Prepare string of Entrez and parse it to qblast\n",
    "    else:\n",
    "        entrez_query = ''\n",
    "        for taxid in list_taxid:\n",
    "            entrez_query += 'txid{}[ORGN]'.format(taxid)\n",
    "            if taxid != list_taxid[-1]:\n",
    "                entrez_query += ' OR '\n",
    "        result_handler = NCBIWWW.qblast('blastp', 'nr', sequence, entrez_query= entrez_query, hitlist_size=query_size)\n",
    "        result_storer = result_handler.read()\n",
    "    logging.info('BLASTp specifying {} taxid(s) completed'.format(len(list_taxid)))\n",
    "\n",
    "    return result_storer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#Single blast query for threading\n",
    "#Threading to be implemented \n",
    "def blastp_single_taxid(sequence, taxid, query_size = 20):\n",
    "    entrez_query = 'txid{}[ORGN]'.format(taxid)\n",
    "    result_handler = NCBIWWW.qblast('blastp', 'nr', sequence, entrez_query= entrez_query, hitlist_size=query_size)    \n",
    "    result_storer = result_handler.read()\n",
    "    return result_storer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#Function to create a temporary file from string and read it with SearchIO.read\n",
    "def xml_string_to_handler(string):\n",
    "    #make temporary file\n",
    "    tmp = tempfile.NamedTemporaryFile(mode='a+')\n",
    "    #write string\n",
    "    tmp.write(string)\n",
    "    handler = SearchIO.read(tmp.name, 'blast-xml')\n",
    "    tmp.close()\n",
    "    logging.info('Blast returned {} results'.format(len(handler)))\n",
    "    return handler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Creation of a dictionary with all HSPS\n",
    "def blast_to_dictionary(blastresult):\n",
    "    blast_dictionary = {'ID' : [], 'Description' : [], 'Seq_length' : [], 'Accession' : [], 'Bitscore' : [], 'Evalue' : [], 'Tot_aln_span':[], 'Identity' :[]}\n",
    "    #Loop through results \n",
    "    for result in blastresult:\n",
    "        blast_dictionary['ID'].append(result.id)\n",
    "        blast_dictionary['Description'].append(result.description)\n",
    "        blast_dictionary['Seq_length'].append(result.seq_len)\n",
    "        blast_dictionary['Accession'].append(result.accession)\n",
    "        #Store results of first HSP\n",
    "        first_hsp = result.hsps[0]\n",
    "        blast_dictionary['Bitscore'].append(first_hsp.bitscore)\n",
    "        blast_dictionary['Evalue'].append(first_hsp.evalue)\n",
    "        #Create variables to store results of multiple hsps\n",
    "        all_alnspan, all_gapnum = [],[] \n",
    "        #Collect data of all hsps for each hit\n",
    "        for hsp in result.hsps:\n",
    "            all_alnspan.append(int(hsp.aln_span))\n",
    "            all_gapnum.append(int(hsp.gap_num))\n",
    "        #Calculate total alignment span and gaps to calculate identity\n",
    "        tot_alnspan, tot_gapnum = int(), int()\n",
    "        seq_len = int(result.seq_len) #DOUBLE CHECK \n",
    "        for span in all_alnspan:\n",
    "            tot_alnspan += span\n",
    "        for gap in all_gapnum:\n",
    "            tot_gapnum += gap\n",
    "        identity = ((tot_alnspan - tot_gapnum)/seq_len)*100\n",
    "        blast_dictionary['Tot_aln_span'].append(tot_alnspan)\n",
    "        blast_dictionary['Identity'].append(round(identity, 3))\n",
    "\n",
    "    logging.info(\"{} entries were recorded from the BLASTp results\".format(len(blast_dictionary['ID'])))\n",
    "    return blast_dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Filter DF based on Evalue, sequence length and identity\n",
    "def filter_df_blast(df:DataFrame, query:SeqRecord, evalue = 10**-10, difference_from_query = 50, identity_threshold=50) -> DataFrame:\n",
    "    #Record initial dataframe length\n",
    "    initial_len = len(df)\n",
    "\n",
    "    remove_index = []\n",
    "    #Check if main HSP is significant for threshold, store indexes of non-significant hits\n",
    "    for i in range(len(df.index)):\n",
    "        eval = df.iloc[i]['Evalue']\n",
    "        if float(eval) > evalue:\n",
    "            remove_index.append(str(i))\n",
    "        \n",
    "    #Copy DF\n",
    "    df_to_return = df    \n",
    "    #Remove indexes if list is not empty\n",
    "    if len(remove_index)>0:\n",
    "        df_to_return = df_to_return.drop(df.index[int(remove_index)])\n",
    "        df_to_return = df_to_return.reset_index(drop=True)\n",
    "    \n",
    "    #Filter by +/- % of query length\n",
    "    #Obtain query length\n",
    "    query_length = len(query.seq)\n",
    "    #Determine upper/lower threshold of acceptance for sequences \n",
    "    lower, upper = query_length*(difference_from_query/100), query_length*((difference_from_query + 100)/100)\n",
    "    #Filter DF\n",
    "    df_to_return = df_to_return[df_to_return['Seq_length'] > lower]\n",
    "    df_to_return = df_to_return.reset_index(drop=True)\n",
    "    df_to_return = df_to_return[df_to_return['Seq_length'] < upper]\n",
    "    df_to_return = df_to_return.reset_index(drop=True)\n",
    "\n",
    "    #Filter by identity\n",
    "    df_to_return = df_to_return[df_to_return['Identity'] > identity_threshold]\n",
    "    df_to_return = df_to_return.reset_index(drop=True)\n",
    "    \n",
    "    #Calculate how many sequences were removed\n",
    "    final_len = len(df_to_return)\n",
    "\n",
    "    logging.info(\"{} sequences were removed filtering the BLAST results DF, returning a DF with {} entries\".format((initial_len-final_len),final_len))\n",
    "    return df_to_return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Retrieve all result's entries from efetch\n",
    "def retrieve_all_efetch(list_of_entries, email):\n",
    "    Bio.Entrez.email = email\n",
    "    list_of_sequences = []\n",
    "    for entry in list_of_entries:\n",
    "        handler = Bio.Entrez.efetch(db='protein', id=entry, rettype = 'fasta',retmode = 'xml', retmax=1) #Returns JSON regardless\n",
    "        gb_info = Bio.Entrez.read(handler, 'text')#Returns nested lists and dictionaries \n",
    "        list_of_sequences.append(gb_info)\n",
    "    logging.info('{} protein entries were retrieved from NCBI protein database'.format(len(list_of_sequences)))\n",
    "    return list_of_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Efetch to dictionary parser\n",
    "def efetch_protein_to_dictionary(list_of_efetch):\n",
    "    #Declare new dictionary\n",
    "    dictionary = {'Accession':[],'Protein_ID':[], 'Taxid':[], 'Organism_name':[], 'Description':[], 'Seq_length':[], 'Prot_sequence':[]}\n",
    "    for wrapper in list_of_efetch:\n",
    "        try:\n",
    "            #Cast into dictionary to avoid random exception\n",
    "            result = dict(wrapper[0])\n",
    "            acc_ver = result['TSeq_accver']\n",
    "            accession = acc_ver.split('.')\n",
    "            dictionary['Accession'].append(accession[0])\n",
    "            dictionary['Protein_ID'].append(result['TSeq_accver'])\n",
    "            dictionary['Taxid'].append(result['TSeq_taxid'])\n",
    "            dictionary['Organism_name'].append(result['TSeq_orgname'])\n",
    "            dictionary['Description'].append(result['TSeq_defline'])\n",
    "            dictionary['Seq_length'].append(result['TSeq_length'])\n",
    "            dictionary['Prot_sequence'].append(result['TSeq_sequence'])\n",
    "        except KeyError:\n",
    "            logging.error('A sequence failed parsing from EFetch')\n",
    "            print('Could not parse one sequence from efetch')\n",
    "    logging.info(\"{} sequences were parsed correctly\".format(len(dictionary['Accession'])))\n",
    "    return dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_df_taxon(df:DataFrame, n_of_sequences = 1) -> DataFrame:\n",
    "    #Record initial length of DF\n",
    "    initial_len = len(df)\n",
    "\n",
    "    #Make a list of all retrieved taxons\n",
    "    retrieved_taxids = df['Taxid'].tolist() \n",
    "    #Make list from dictionary to eliminate duplicates \n",
    "    retrieved_taxids = list(dict.fromkeys(retrieved_taxids))\n",
    "        \n",
    "    #Declare empty DF\n",
    "    df_toreturn = pd.DataFrame()\n",
    "\n",
    "    #For each taxon create a DF, sort and get best result to append to df_toreturn\n",
    "    for taxid in retrieved_taxids:\n",
    "        #Create temporary DF exclusive to taxon\n",
    "        temp_df = df[df['Taxid'] == taxid]\n",
    "        temp_df = temp_df.sort_values(['Evalue', 'Identity', 'Bitscore'], ascending=[True, False, False]) #('Identity', ascending=False)\n",
    "        if len(temp_df) > 0:\n",
    "            if len(temp_df) > n_of_sequences:\n",
    "                df_toreturn = df_toreturn.append(temp_df[:n_of_sequences])\n",
    "            else:\n",
    "                df_toreturn = df_toreturn.append(temp_df)\n",
    "\n",
    "    #Refactor indexes\n",
    "    df_toreturn = df_toreturn.reset_index(drop=True)\n",
    "\n",
    "    #Calculate final length and append info to logging \n",
    "    final_len = len(df_toreturn)\n",
    "    logging.info(\"Filtering by taxon: {} unique taxid(s) collected, {} entries discarded\".format(len(retrieved_taxids), (initial_len - final_len)))\n",
    "    return df_toreturn\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare string for alingment file\n",
    "def fasta_for_alignment(query:SeqRecord, df:DataFrame) -> str:\n",
    "    #Initialise string and add \n",
    "    string = ''\n",
    "    if query.id != '<unknown id>':\n",
    "        string += \">Query:{}\\n\".format(query.id)\n",
    "    else:\n",
    "        string += \">Query\\n\"\n",
    "    string += \"{}\\n\".format(query.seq)\n",
    "    #Add all sequence IDs and aa sequence\n",
    "    for i in range(len(df['Accession'])):\n",
    "        #Check if protein is missing ------------------------------------------------Add to report? ERROR.LOG FILE!!!!!\n",
    "        if len(df['Prot_sequence'][i]) <1:\n",
    "            continue\n",
    "        string += \">{}\\n\".format(df['Protein_ID'][i])#, df['Organism_name'][i], df['Taxid'][i], df['Description'][i])\n",
    "        string += \"{}\".format(df['Prot_sequence'][i])\n",
    "        #Check if is not the last element in the list \n",
    "        if i != len(df['Accession']):\n",
    "            string += '\\n'\n",
    "    logging.info(\"{} sequences were prepared for alignment with the query\".format(len(df)))\n",
    "    return string\n",
    "\n",
    "\n",
    "\n",
    "#Change 'defining string' from protein to protein name/species name/taxid!!!!!!!!!!!!!!\n",
    "#For accession replace accession + name!!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run mafft by saving the fasta sequences for alignment in a file and passing it to mafft \n",
    "def run_mafft_saving_file(fasta:str, mafft_directory:str, filename:str) -> str:\n",
    "    #Write fasta file for alignment\n",
    "    file = '{}.fasta'.format(filename)\n",
    "    with open(file, 'w') as handle:\n",
    "        handle.write(fasta)\n",
    "    #Parse and run with mafft \n",
    "    command_list = [mafft_directory, '--distout', '{}'.format(file)]\n",
    "    process = subprocess.Popen(command_list, universal_newlines= True, stdout = subprocess.PIPE, stderr = subprocess.PIPE)\n",
    "    stdout, stderr = process.communicate()\n",
    "\n",
    "    #Save standard error of mafft\n",
    "    with open('aligned_mafft.stderr', 'w') as handle:\n",
    "        handle.write(stderr)\n",
    "    #Retrun alignment \n",
    "    return stdout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to get protein sequence and save fasta file given an accession number\n",
    "def get_fasta_from_accession(accession, email):\n",
    "\n",
    "    #Get the efetch handler \n",
    "    Bio.Entrez.email = email\n",
    "    handler = Bio.Entrez.efetch(db='protein', id=accession, rettype = 'fasta',retmode = 'xml', retmax=1) #Returns JSON regardless\n",
    "    query_protein_efetch = Bio.Entrez.read(handler, 'text')#Returns nested lists and dictionaries \n",
    "    \n",
    "    #Make a dictionary -- passed as list to recycle the efetch_protein_to_dictionary function\n",
    "    dictionary_query = efetch_protein_to_dictionary([query_protein_efetch])\n",
    "\n",
    "    #Make and save fasta file \n",
    "    fasta_string_query = \">{} \\n{}\".format(dictionary_query['Accession'][0], dictionary_query['Prot_sequence'][0])\n",
    "\n",
    "    fasta_file_name = \"{}_sequence.fasta\".format(dictionary_query['Accession'][0])\n",
    "    with open(fasta_file_name, 'w') as handle:\n",
    "        handle.write(fasta_string_query)\n",
    "\n",
    "    #Open sequence with open_fasta and return it \n",
    "    fasta_record_q = open_fasta(fasta_file_name)\n",
    "    return fasta_record_q"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def open_input(input, email):\n",
    "    #Declare empty protein sequence\n",
    "    protein_sequence = None\n",
    "    #Check that the parsed string has a fasta extension, if it does pass it to open fasta function\n",
    "    if input.endswith('.fas') or input.endswith('.fasta'):\n",
    "        sequence = open_fasta(input)\n",
    "        #Try translating the sequence, if an error is raised the sequence is already a peptide and can be returned\n",
    "        try: \n",
    "            protein_sequence = sequence.translate(to_stop = True)\n",
    "            logging.info('Nucleotide sequence was translated to protein')\n",
    "        except Exception:\n",
    "            protein_sequence = sequence\n",
    "            logging.info('Protein sequence was opened')\n",
    "    #If the string doesn't have an extension it will be passed to the get_fasta_from_accession function\n",
    "    else:\n",
    "        try:\n",
    "            protein_sequence = get_fasta_from_accession(input, email)\n",
    "            logging.info('Protein sequence was retrieved from NCBI protein database')\n",
    "        except Exception as e:\n",
    "            logging.error(traceback.format_exc())\n",
    "            logging.error('The file must have a .fas or .fasta extension')\n",
    "\n",
    "    return protein_sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tree_from_alignment(alignment):\n",
    "    calculator = DistanceCalculator('identity')\n",
    "    distance_matrix = calculator.get_distance(alignment)\n",
    "\n",
    "    constructor = DistanceTreeConstructor(calculator, 'nj')\n",
    "    tree = constructor.build_tree(alignment)\n",
    "    logging.info('Tree was produced with this and that method') #Placeholder to change\n",
    "    return tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#USER INPUTS\n",
    "input_string = 'human_mx1.fas' #'sry_protein.fasta' 'QBA69874'\n",
    "taxid_list = [] #['9592']\n",
    "mafft_directory = r'/Users/Gioele/miniconda3/bin/mafft'\n",
    "email = 'A.N.Other@example.com'\n",
    "output_name = 'draft_v4_debug_3'\n",
    "\n",
    "\n",
    "local_query = False\n",
    "threading = False\n",
    "query_size = 100\n",
    "\n",
    "evalue_threshold = 10**-10\n",
    "len_threshold = 50\n",
    "identity_threshold = 50\n",
    "\n",
    "sequences_per_taxon = 1\n",
    "\n",
    "#Make tsv for figtree compatibility\n",
    "make_tsv = False\n",
    "\n",
    "#logging file \n",
    "logging.basicConfig(filename='{}.log'.format(output_name), filemode='w', format='%(levelname)s:%(message)s', level=logging.DEBUG) #logging refreshes every run and only displays type of message: message\n",
    "\n",
    "#Filenames:\n",
    "#Implemented\n",
    "output_df = '{}_df.csv'.format(output_name)\n",
    "output_alignment = '{}_alignment.fasta'.format(output_name)\n",
    "output_xml_tree = '{}_xml_tree.xml'.format(output_name)\n",
    "#To add\n",
    "output_tree_newick = '{}_newick_tree.nwk'.format(output_name) #!!!!\n",
    "output_tree_jpg = '{}_tree_image.jpg'.format(output_name)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#Open fasta record\n",
    "fasta_record = open_input(input_string, email)\n",
    "#Blast with list \n",
    "blast_results = blastp_with_list(fasta_record.seq, query_size=query_size, list_taxid=taxid_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Blast w threading \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BLAST DF after filtering holds 100 entries\n"
     ]
    }
   ],
   "source": [
    "#!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!NEED TO DOUBLECHECK AND IMPLEMENT BLAST WITH THREADING \n",
    "\n",
    "\n",
    "#Parse handler to xml_string_to_handler to be able to feed it in SearchIO.read\n",
    "handler_blast = xml_string_to_handler(blast_results)\n",
    "\n",
    "#Make dictionary from handler\n",
    "dictionary_blast = blast_to_dictionary(handler_blast)\n",
    "\n",
    "#Transform dictionary into DF\n",
    "results_df_blast = pd.DataFrame.from_dict(dictionary_blast)\n",
    "\n",
    "\n",
    "#Filter DF by E-value / Bitscore / Identity\n",
    "filtered_blast_df = filter_df_blast(results_df_blast, fasta_record, evalue=evalue_threshold, difference_from_query=len_threshold, identity_threshold=identity_threshold)\n",
    "\n",
    "print(\"BLAST DF after filtering holds {} entries\".format(len(results_df_blast)))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could not parse one sequence from efetch\n",
      "Could not parse one sequence from efetch\n",
      "Filtered DF presents 23 entries\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#Retrieve all results with efetch \n",
    "#Retrieve accession list\n",
    "accession_list = filtered_blast_df['Accession'].tolist()\n",
    "#Get results \n",
    "efetch_result = retrieve_all_efetch(accession_list, email)\n",
    "\n",
    "# Make dictionary from efetch fasta results\n",
    "dictionary_efetch = efetch_protein_to_dictionary(efetch_result)\n",
    "\n",
    "#Transform dictionary into DF\n",
    "efetch_df = pd.DataFrame.from_dict(dictionary_efetch)\n",
    "\n",
    "\n",
    "\n",
    "#Merge blast and efetch DataFrames\n",
    "left = filtered_blast_df.loc[:,['Accession', 'Seq_length', 'Evalue', 'Bitscore', 'Tot_aln_span', 'Identity']]\n",
    "right = efetch_df.loc[:,['Accession', 'Protein_ID' ,'Taxid', 'Organism_name', 'Description', 'Prot_sequence']]\n",
    "combined_df = pd.merge(left, right, on='Accession')\n",
    "\n",
    "\n",
    "#Filter DF based on taxons\n",
    "filtered_df = filter_df_taxon(combined_df, n_of_sequences=sequences_per_taxon)\n",
    "\n",
    "\n",
    "#Save DF of sequences that are going to be aligned \n",
    "filtered_df.to_csv(output_df, index = False)\n",
    "\n",
    "\n",
    "print('Filtered DF presents {} entries'.format(len(output_df)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make TSV for figtree\n",
    "tsv_df = filtered_df[['Protein_ID'] + [col for col in filtered_df.columns if col!= 'Protein_ID']]\n",
    "tsv_df = tsv_df.drop('Prot_sequence', axis = 1)\n",
    "tsv_df.to_csv('{}.tsv'.format(output_df), sep = '\\t', index = False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Prepare for multiple alingment\n",
    "fasta_string_for_alignment = fasta_for_alignment(fasta_record, filtered_df)\n",
    "\n",
    "#Run mafft alignment saving file\n",
    "mafft_alignment = run_mafft_saving_file(fasta_string_for_alignment, mafft_directory, 'multiple_seq_fasta')\n",
    "\n",
    "#Save mafft alignment\n",
    "with open(output_alignment, 'w') as savefile:\n",
    "    savefile.write(mafft_alignment)\n",
    "\n",
    "\n",
    "#Open AlignIO from fasta file \n",
    "alignment = AlignIO.read(output_alignment, 'fasta')\n",
    "\n",
    "#Construct tree from alignment \n",
    "#To IMPLEMENT DIFFERENT METHODS FOR TREE BUILDING \n",
    "tree = tree_from_alignment(alignment)  \n",
    "\n",
    "\n",
    "#Write tree in xml\n",
    "Phylo.write(tree, output_xml_tree, 'phyloxml')\n",
    "\n",
    "Phylo.write(tree, output_tree_newick, 'newick')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Re import newick string to feed to ETE3\n",
    "with open(output_tree_newick, 'r') as handle:\n",
    "    newick_string = handle.read()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'Organism_name': 'Mustela putorius furo', 'Protein_ID': 'XP_004762249.1'}, {'Organism_name': 'Nyctereutes procyonoides', 'Protein_ID': 'CAD7693676.1'}, {'Organism_name': 'Chlorocebus sabaeus', 'Protein_ID': 'XP_007966881.2'}, {'Organism_name': 'Suricata suricatta', 'Protein_ID': 'XP_029794794.1'}, {'Organism_name': 'Puma yagouaroundi', 'Protein_ID': 'XP_040339814.1'}, {'Organism_name': 'Mustela erminea', 'Protein_ID': 'XP_032211404.1'}, {'Organism_name': 'Propithecus coquereli', 'Protein_ID': 'XP_012495571.1'}, {'Organism_name': 'Lynx canadensis', 'Protein_ID': 'XP_030185366.1'}, {'Organism_name': 'Felis catus', 'Protein_ID': 'XP_023094481.1'}, {'Organism_name': 'Acinonyx jubatus', 'Protein_ID': 'XP_026897378.1'}, {'Organism_name': 'Vulpes vulpes', 'Protein_ID': 'XP_025839180.1'}, {'Organism_name': 'Miopithecus talapoin', 'Protein_ID': 'AFU81311.1'}, {'Organism_name': 'Alouatta sara', 'Protein_ID': 'AFU81300.1'}, {'Organism_name': 'Sus scrofa', 'Protein_ID': 'BAD11809.1'}, {'Organism_name': 'Macaca nemestrina', 'Protein_ID': 'XP_011724293.1'}, {'Organism_name': 'Sapajus apella', 'Protein_ID': 'XP_032132720.1'}, {'Organism_name': 'Prionailurus bengalensis', 'Protein_ID': 'XP_043449495.1'}, {'Organism_name': 'Vulpes lagopus', 'Protein_ID': 'XP_041591175.1'}, {'Organism_name': 'Panthera leo', 'Protein_ID': 'XP_042811978.1'}, {'Organism_name': 'Orycteropus afer afer', 'Protein_ID': 'XP_007945509.1'}, {'Organism_name': 'Lagothrix lagotricha', 'Protein_ID': 'AFU81308.1'}, {'Organism_name': 'Macaca sylvanus', 'Protein_ID': 'AFU81310.1'}, {'Organism_name': 'Nomascus leucogenys', 'Protein_ID': 'XP_012357180.1'}, {'Organism_name': 'Pithecia pithecia', 'Protein_ID': 'AFU81315.1'}, {'Organism_name': 'Loxodonta africana', 'Protein_ID': 'XP_010596962.1'}, {'Organism_name': 'synthetic construct', 'Protein_ID': 'AAX36236.1'}, {'Organism_name': 'Lynx pardinus', 'Protein_ID': 'VFV35827.1'}, {'Organism_name': 'Manis javanica', 'Protein_ID': 'XP_036870729.1'}, {'Organism_name': 'Manis pentadactyla', 'Protein_ID': 'XP_036732799.1'}, {'Organism_name': 'Hyaena hyaena', 'Protein_ID': 'XP_039085427.1'}, {'Organism_name': 'Saimiri sciureus', 'Protein_ID': 'AFU81318.1'}, {'Organism_name': 'Hylobates agilis', 'Protein_ID': 'AFU81307.1'}, {'Organism_name': 'Carlito syrichta', 'Protein_ID': 'XP_008066755.1'}, {'Organism_name': 'Eptesicus fuscus', 'Protein_ID': 'XP_008143913.1'}, {'Organism_name': 'Saguinus midas', 'Protein_ID': 'AFU81317.1'}, {'Organism_name': 'Pan paniscus', 'Protein_ID': 'NP_001288691.1'}, {'Organism_name': 'Ceratotherium simum simum', 'Protein_ID': 'XP_004429677.1'}, {'Organism_name': 'Callithrix jacchus', 'Protein_ID': 'XP_008984967.1'}, {'Organism_name': 'Panthera tigris', 'Protein_ID': 'XP_042812427.1'}, {'Organism_name': 'Colobus angolensis palliatus', 'Protein_ID': 'XP_011809449.1'}, {'Organism_name': 'Pongo abelii', 'Protein_ID': 'PNJ42385.1'}, {'Organism_name': 'Macaca fascicularis', 'Protein_ID': 'AFU81309.1'}, {'Organism_name': 'Hylobates moloch', 'Protein_ID': 'XP_032030149.1'}, {'Organism_name': 'Papio anubis', 'Protein_ID': 'XP_003895511.1'}, {'Organism_name': 'Plecturocebus moloch', 'Protein_ID': 'AFU81302.1'}, {'Organism_name': 'Pan troglodytes', 'Protein_ID': 'XP_009431729.1'}, {'Organism_name': 'Colobus guereza', 'Protein_ID': 'AFU81305.1'}, {'Organism_name': 'Mandrillus leucophaeus', 'Protein_ID': 'XP_011839575.1'}, {'Organism_name': 'Piliocolobus tephrosceles', 'Protein_ID': 'XP_023047606.1'}, {'Organism_name': 'Rhinopithecus roxellana', 'Protein_ID': 'XP_030770463.1'}, {'Organism_name': 'Galeopterus variegatus', 'Protein_ID': 'XP_008569442.1'}, {'Organism_name': 'Tupaia chinensis', 'Protein_ID': 'XP_006156437.1'}, {'Organism_name': 'Cercocebus atys', 'Protein_ID': 'XP_011892717.1'}, {'Organism_name': 'Dasypus novemcinctus', 'Protein_ID': 'XP_004466363.1'}, {'Organism_name': 'Balaenoptera acutorostrata scammoni', 'Protein_ID': 'XP_007172774.1'}, {'Organism_name': 'Aotus nancymaae', 'Protein_ID': 'XP_012300528.1'}, {'Organism_name': 'Lontra canadensis', 'Protein_ID': 'XP_032721188.1'}, {'Organism_name': 'Macaca mulatta', 'Protein_ID': 'XP_014988323.2'}, {'Organism_name': 'Gorilla gorilla', 'Protein_ID': 'NP_001266690.1'}, {'Organism_name': 'Theropithecus gelada', 'Protein_ID': 'XP_025236083.1'}, {'Organism_name': 'Homo sapiens', 'Protein_ID': 'NP_001138397.1'}, {'Organism_name': 'Aotus trivirgatus', 'Protein_ID': 'AFU81301.1'}, {'Organism_name': 'Saimiri boliviensis boliviensis', 'Protein_ID': 'XP_003927641.1'}, {'Organism_name': 'Pongo pygmaeus', 'Protein_ID': 'AFU81316.1'}, {'Organism_name': 'Chlorocebus tantalus', 'Protein_ID': 'AFU81304.1'}, {'Organism_name': 'Cebus imitator', 'Protein_ID': 'XP_037597997.1'}, {'Organism_name': 'Trachypithecus francoisi', 'Protein_ID': 'AFU81320.1'}, {'Organism_name': 'Rhinopithecus bieti', 'Protein_ID': 'XP_017735713.1'}]\n"
     ]
    }
   ],
   "source": [
    "#Create a DF with only protein ids and organisms names and drop the index\n",
    "dict_of_df = filtered_df[['Protein_ID', 'Organism_name']]\n",
    "dict_of_df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "#Create a dictionary from DF\n",
    "dict_of_df = dict_of_df.to_dict('records')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Replace all organisms names' spaces with underscores to avoid confusion\n",
    "for entry in dict_of_df:\n",
    "    entry['Organism_name'] = str(entry['Organism_name']).replace(' ', '_')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'Organism_name': 'Mustela_putorius_furo', 'Protein_ID': 'XP_004762249.1'},\n",
       " {'Organism_name': 'Nyctereutes_procyonoides', 'Protein_ID': 'CAD7693676.1'},\n",
       " {'Organism_name': 'Chlorocebus_sabaeus', 'Protein_ID': 'XP_007966881.2'},\n",
       " {'Organism_name': 'Suricata_suricatta', 'Protein_ID': 'XP_029794794.1'},\n",
       " {'Organism_name': 'Puma_yagouaroundi', 'Protein_ID': 'XP_040339814.1'},\n",
       " {'Organism_name': 'Mustela_erminea', 'Protein_ID': 'XP_032211404.1'},\n",
       " {'Organism_name': 'Propithecus_coquereli', 'Protein_ID': 'XP_012495571.1'},\n",
       " {'Organism_name': 'Lynx_canadensis', 'Protein_ID': 'XP_030185366.1'},\n",
       " {'Organism_name': 'Felis_catus', 'Protein_ID': 'XP_023094481.1'},\n",
       " {'Organism_name': 'Acinonyx_jubatus', 'Protein_ID': 'XP_026897378.1'},\n",
       " {'Organism_name': 'Vulpes_vulpes', 'Protein_ID': 'XP_025839180.1'},\n",
       " {'Organism_name': 'Miopithecus_talapoin', 'Protein_ID': 'AFU81311.1'},\n",
       " {'Organism_name': 'Alouatta_sara', 'Protein_ID': 'AFU81300.1'},\n",
       " {'Organism_name': 'Sus_scrofa', 'Protein_ID': 'BAD11809.1'},\n",
       " {'Organism_name': 'Macaca_nemestrina', 'Protein_ID': 'XP_011724293.1'},\n",
       " {'Organism_name': 'Sapajus_apella', 'Protein_ID': 'XP_032132720.1'},\n",
       " {'Organism_name': 'Prionailurus_bengalensis', 'Protein_ID': 'XP_043449495.1'},\n",
       " {'Organism_name': 'Vulpes_lagopus', 'Protein_ID': 'XP_041591175.1'},\n",
       " {'Organism_name': 'Panthera_leo', 'Protein_ID': 'XP_042811978.1'},\n",
       " {'Organism_name': 'Orycteropus_afer_afer', 'Protein_ID': 'XP_007945509.1'},\n",
       " {'Organism_name': 'Lagothrix_lagotricha', 'Protein_ID': 'AFU81308.1'},\n",
       " {'Organism_name': 'Macaca_sylvanus', 'Protein_ID': 'AFU81310.1'},\n",
       " {'Organism_name': 'Nomascus_leucogenys', 'Protein_ID': 'XP_012357180.1'},\n",
       " {'Organism_name': 'Pithecia_pithecia', 'Protein_ID': 'AFU81315.1'},\n",
       " {'Organism_name': 'Loxodonta_africana', 'Protein_ID': 'XP_010596962.1'},\n",
       " {'Organism_name': 'synthetic_construct', 'Protein_ID': 'AAX36236.1'},\n",
       " {'Organism_name': 'Lynx_pardinus', 'Protein_ID': 'VFV35827.1'},\n",
       " {'Organism_name': 'Manis_javanica', 'Protein_ID': 'XP_036870729.1'},\n",
       " {'Organism_name': 'Manis_pentadactyla', 'Protein_ID': 'XP_036732799.1'},\n",
       " {'Organism_name': 'Hyaena_hyaena', 'Protein_ID': 'XP_039085427.1'},\n",
       " {'Organism_name': 'Saimiri_sciureus', 'Protein_ID': 'AFU81318.1'},\n",
       " {'Organism_name': 'Hylobates_agilis', 'Protein_ID': 'AFU81307.1'},\n",
       " {'Organism_name': 'Carlito_syrichta', 'Protein_ID': 'XP_008066755.1'},\n",
       " {'Organism_name': 'Eptesicus_fuscus', 'Protein_ID': 'XP_008143913.1'},\n",
       " {'Organism_name': 'Saguinus_midas', 'Protein_ID': 'AFU81317.1'},\n",
       " {'Organism_name': 'Pan_paniscus', 'Protein_ID': 'NP_001288691.1'},\n",
       " {'Organism_name': 'Ceratotherium_simum_simum',\n",
       "  'Protein_ID': 'XP_004429677.1'},\n",
       " {'Organism_name': 'Callithrix_jacchus', 'Protein_ID': 'XP_008984967.1'},\n",
       " {'Organism_name': 'Panthera_tigris', 'Protein_ID': 'XP_042812427.1'},\n",
       " {'Organism_name': 'Colobus_angolensis_palliatus',\n",
       "  'Protein_ID': 'XP_011809449.1'},\n",
       " {'Organism_name': 'Pongo_abelii', 'Protein_ID': 'PNJ42385.1'},\n",
       " {'Organism_name': 'Macaca_fascicularis', 'Protein_ID': 'AFU81309.1'},\n",
       " {'Organism_name': 'Hylobates_moloch', 'Protein_ID': 'XP_032030149.1'},\n",
       " {'Organism_name': 'Papio_anubis', 'Protein_ID': 'XP_003895511.1'},\n",
       " {'Organism_name': 'Plecturocebus_moloch', 'Protein_ID': 'AFU81302.1'},\n",
       " {'Organism_name': 'Pan_troglodytes', 'Protein_ID': 'XP_009431729.1'},\n",
       " {'Organism_name': 'Colobus_guereza', 'Protein_ID': 'AFU81305.1'},\n",
       " {'Organism_name': 'Mandrillus_leucophaeus', 'Protein_ID': 'XP_011839575.1'},\n",
       " {'Organism_name': 'Piliocolobus_tephrosceles',\n",
       "  'Protein_ID': 'XP_023047606.1'},\n",
       " {'Organism_name': 'Rhinopithecus_roxellana', 'Protein_ID': 'XP_030770463.1'},\n",
       " {'Organism_name': 'Galeopterus_variegatus', 'Protein_ID': 'XP_008569442.1'},\n",
       " {'Organism_name': 'Tupaia_chinensis', 'Protein_ID': 'XP_006156437.1'},\n",
       " {'Organism_name': 'Cercocebus_atys', 'Protein_ID': 'XP_011892717.1'},\n",
       " {'Organism_name': 'Dasypus_novemcinctus', 'Protein_ID': 'XP_004466363.1'},\n",
       " {'Organism_name': 'Balaenoptera_acutorostrata_scammoni',\n",
       "  'Protein_ID': 'XP_007172774.1'},\n",
       " {'Organism_name': 'Aotus_nancymaae', 'Protein_ID': 'XP_012300528.1'},\n",
       " {'Organism_name': 'Lontra_canadensis', 'Protein_ID': 'XP_032721188.1'},\n",
       " {'Organism_name': 'Macaca_mulatta', 'Protein_ID': 'XP_014988323.2'},\n",
       " {'Organism_name': 'Gorilla_gorilla', 'Protein_ID': 'NP_001266690.1'},\n",
       " {'Organism_name': 'Theropithecus_gelada', 'Protein_ID': 'XP_025236083.1'},\n",
       " {'Organism_name': 'Homo_sapiens', 'Protein_ID': 'NP_001138397.1'},\n",
       " {'Organism_name': 'Aotus_trivirgatus', 'Protein_ID': 'AFU81301.1'},\n",
       " {'Organism_name': 'Saimiri_boliviensis_boliviensis',\n",
       "  'Protein_ID': 'XP_003927641.1'},\n",
       " {'Organism_name': 'Pongo_pygmaeus', 'Protein_ID': 'AFU81316.1'},\n",
       " {'Organism_name': 'Chlorocebus_tantalus', 'Protein_ID': 'AFU81304.1'},\n",
       " {'Organism_name': 'Cebus_imitator', 'Protein_ID': 'XP_037597997.1'},\n",
       " {'Organism_name': 'Trachypithecus_francoisi', 'Protein_ID': 'AFU81320.1'},\n",
       " {'Organism_name': 'Rhinopithecus_bieti', 'Protein_ID': 'XP_017735713.1'}]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict_of_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mustela_putorius_furo|XP_004762249.1\n",
      "Nyctereutes_procyonoides|CAD7693676.1\n",
      "Chlorocebus_sabaeus|XP_007966881.2\n",
      "Suricata_suricatta|XP_029794794.1\n",
      "Puma_yagouaroundi|XP_040339814.1\n",
      "Mustela_erminea|XP_032211404.1\n",
      "Propithecus_coquereli|XP_012495571.1\n",
      "Lynx_canadensis|XP_030185366.1\n",
      "Felis_catus|XP_023094481.1\n",
      "Acinonyx_jubatus|XP_026897378.1\n",
      "Vulpes_vulpes|XP_025839180.1\n",
      "Miopithecus_talapoin|AFU81311.1\n",
      "Alouatta_sara|AFU81300.1\n",
      "Sus_scrofa|BAD11809.1\n",
      "Macaca_nemestrina|XP_011724293.1\n",
      "Sapajus_apella|XP_032132720.1\n",
      "Prionailurus_bengalensis|XP_043449495.1\n",
      "Vulpes_lagopus|XP_041591175.1\n",
      "Panthera_leo|XP_042811978.1\n",
      "Orycteropus_afer_afer|XP_007945509.1\n",
      "Lagothrix_lagotricha|AFU81308.1\n",
      "Macaca_sylvanus|AFU81310.1\n",
      "Nomascus_leucogenys|XP_012357180.1\n",
      "Pithecia_pithecia|AFU81315.1\n",
      "Loxodonta_africana|XP_010596962.1\n",
      "synthetic_construct|AAX36236.1\n",
      "Lynx_pardinus|VFV35827.1\n",
      "Manis_javanica|XP_036870729.1\n",
      "Manis_pentadactyla|XP_036732799.1\n",
      "Hyaena_hyaena|XP_039085427.1\n",
      "Saimiri_sciureus|AFU81318.1\n",
      "Hylobates_agilis|AFU81307.1\n",
      "Carlito_syrichta|XP_008066755.1\n",
      "Eptesicus_fuscus|XP_008143913.1\n",
      "Saguinus_midas|AFU81317.1\n",
      "Pan_paniscus|NP_001288691.1\n",
      "Ceratotherium_simum_simum|XP_004429677.1\n",
      "Callithrix_jacchus|XP_008984967.1\n",
      "Panthera_tigris|XP_042812427.1\n",
      "Colobus_angolensis_palliatus|XP_011809449.1\n",
      "Pongo_abelii|PNJ42385.1\n",
      "Macaca_fascicularis|AFU81309.1\n",
      "Hylobates_moloch|XP_032030149.1\n",
      "Papio_anubis|XP_003895511.1\n",
      "Plecturocebus_moloch|AFU81302.1\n",
      "Pan_troglodytes|XP_009431729.1\n",
      "Colobus_guereza|AFU81305.1\n",
      "Mandrillus_leucophaeus|XP_011839575.1\n",
      "Piliocolobus_tephrosceles|XP_023047606.1\n",
      "Rhinopithecus_roxellana|XP_030770463.1\n",
      "Galeopterus_variegatus|XP_008569442.1\n",
      "Tupaia_chinensis|XP_006156437.1\n",
      "Cercocebus_atys|XP_011892717.1\n",
      "Dasypus_novemcinctus|XP_004466363.1\n",
      "Balaenoptera_acutorostrata_scammoni|XP_007172774.1\n",
      "Aotus_nancymaae|XP_012300528.1\n",
      "Lontra_canadensis|XP_032721188.1\n",
      "Macaca_mulatta|XP_014988323.2\n",
      "Gorilla_gorilla|NP_001266690.1\n",
      "Theropithecus_gelada|XP_025236083.1\n",
      "Homo_sapiens|NP_001138397.1\n",
      "Aotus_trivirgatus|AFU81301.1\n",
      "Saimiri_boliviensis_boliviensis|XP_003927641.1\n",
      "Pongo_pygmaeus|AFU81316.1\n",
      "Chlorocebus_tantalus|AFU81304.1\n",
      "Cebus_imitator|XP_037597997.1\n",
      "Trachypithecus_francoisi|AFU81320.1\n",
      "Rhinopithecus_bieti|XP_017735713.1\n"
     ]
    }
   ],
   "source": [
    "#Change all instances of protein ID to Organism name | protein ID in both newick and MSA strings\n",
    "#Used to make the ETE3 output more human readable\n",
    "for entry in dict_of_df:\n",
    "    pro_id = str(entry['Protein_ID'])\n",
    "    org_name = str(entry['Organism_name'])\n",
    "    new_entry = '{}|{}'.format(org_name, pro_id)\n",
    "    print(new_entry)\n",
    "\n",
    "    newick_string= newick_string.replace(pro_id, new_entry)\n",
    "    mafft_alignment= mafft_alignment.replace(pro_id, new_entry)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "draft_v4_debug_3_alignment.fasta\n"
     ]
    }
   ],
   "source": [
    "#print(newick_string)\n",
    "\n",
    "print(output_alignment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"leaf_x = None\\nfor leaf in msa_tree.iter_leaves():\\n    print('Name: {}, Sequence: {}'.format(leaf.name, leaf.sequence))\\n    leaf_x=leaf\""
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from ete3 import PhyloTree, TreeStyle\n",
    "\n",
    "#Link tree with MSA\n",
    "\n",
    "msa_tree = PhyloTree(newick_string, quoted_node_names=True, format=1) \n",
    "#Need to root the tree!!!!!!!\n",
    "#Returns the node that divides the current tree into two distance-balanced partitions.\n",
    "R = msa_tree.get_midpoint_outgroup()\n",
    "#Sets a descendant node as the outgroup of a tree\n",
    "msa_tree.set_outgroup(R)\n",
    "\n",
    "#Link tree to the MAFFT Alignment \n",
    "msa_tree.link_to_alignment(alignment=mafft_alignment, alg_format='fasta')\n",
    "msa_tree\n",
    "\n",
    "'''leaf_x = None\n",
    "for leaf in msa_tree.iter_leaves():\n",
    "    print('Name: {}, Sequence: {}'.format(leaf.name, leaf.sequence))\n",
    "    leaf_x=leaf'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "msa_tree.show(tree_style=TreeStyle())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'faces': [[175.1005898789196,\n",
       "   859.0,\n",
       "   339.1005898789196,\n",
       "   872.0,\n",
       "   133,\n",
       "   'Macaca_fascicularis|AFU81309.1'],\n",
       "  [452.907947842285, 858.5, 453.907947842285, 872.5, 133, None],\n",
       "  [149.81403290903444,\n",
       "   131.0,\n",
       "   309.81403290903444,\n",
       "   144.0,\n",
       "   23,\n",
       "   'Vulpes_vulpes|XP_025839180.1'],\n",
       "  [452.907947842285, 130.5, 453.907947842285, 144.5, 23, None],\n",
       "  [102.19031356721514,\n",
       "   105.0,\n",
       "   270.1903135672151,\n",
       "   118.0,\n",
       "   19,\n",
       "   'Hyaena_hyaena|XP_039085427.1'],\n",
       "  [452.907947842285, 104.5, 453.907947842285, 118.5, 19, None],\n",
       "  [109.62278795405153,\n",
       "   118.0,\n",
       "   285.62278795405155,\n",
       "   131.0,\n",
       "   20,\n",
       "   'Suricata_suricatta|XP_029794794.1'],\n",
       "  [452.907947842285, 117.5, 453.907947842285, 131.5, 20, None],\n",
       "  [182.58708475628688,\n",
       "   833.0,\n",
       "   336.5870847562869,\n",
       "   846.0,\n",
       "   130,\n",
       "   'Papio_anubis|XP_003895511.1'],\n",
       "  [452.907947842285, 832.5, 453.907947842285, 846.5, 130, None],\n",
       "  [181.07016454517228,\n",
       "   443.0,\n",
       "   416.07016454517225,\n",
       "   456.0,\n",
       "   72,\n",
       "   'Saimiri_boliviensis_boliviensis|XP_003927641.1'],\n",
       "  [452.907947842285, 442.5, 453.907947842285, 456.5, 72, None],\n",
       "  [167.43340577460413,\n",
       "   417.0,\n",
       "   314.4334057746041,\n",
       "   430.0,\n",
       "   68,\n",
       "   'Aotus_trivirgatus|AFU81301.1'],\n",
       "  [452.907947842285, 416.5, 453.907947842285, 430.5, 68, None],\n",
       "  [177.69295249922385,\n",
       "   794.0,\n",
       "   349.69295249922385,\n",
       "   807.0,\n",
       "   125,\n",
       "   'Cercocebus_atys|XP_011892717.1'],\n",
       "  [452.907947842285, 793.5, 453.907947842285, 807.5, 125, None],\n",
       "  [171.7258615336852,\n",
       "   781.0,\n",
       "   369.7258615336852,\n",
       "   794.0,\n",
       "   121,\n",
       "   'Chlorocebus_sabaeus|XP_007966881.2'],\n",
       "  [452.907947842285, 780.5, 453.907947842285, 794.5, 121, None],\n",
       "  [168.52995963986336,\n",
       "   469.0,\n",
       "   330.52995963986336,\n",
       "   482.0,\n",
       "   75,\n",
       "   'Cebus_imitator|XP_037597997.1'],\n",
       "  [452.907947842285, 468.5, 453.907947842285, 482.5, 75, None],\n",
       "  [95.97547345544862,\n",
       "   144.0,\n",
       "   260.97547345544865,\n",
       "   157.0,\n",
       "   25,\n",
       "   'Vulpes_lagopus|XP_041591175.1'],\n",
       "  [452.907947842285, 143.5, 453.907947842285, 157.5, 25, None],\n",
       "  [144.19093449239364,\n",
       "   287.0,\n",
       "   318.19093449239364,\n",
       "   300.0,\n",
       "   46,\n",
       "   'Tupaia_chinensis|XP_006156437.1'],\n",
       "  [452.907947842285, 286.5, 453.907947842285, 300.5, 46, None],\n",
       "  [193.0, 274.0, 367.0, 287.0, 43, 'Eptesicus_fuscus|XP_008143913.1'],\n",
       "  [452.907947842285, 273.5, 453.907947842285, 287.5, 43, None],\n",
       "  [174.43371623719338,\n",
       "   313.0,\n",
       "   359.4337162371934,\n",
       "   326.0,\n",
       "   50,\n",
       "   'Loxodonta_africana|XP_010596962.1'],\n",
       "  [452.907947842285, 312.5, 453.907947842285, 326.5, 50, None],\n",
       "  [182.5007761564731,\n",
       "   456.0,\n",
       "   328.5007761564731,\n",
       "   469.0,\n",
       "   73,\n",
       "   'Saimiri_sciureus|AFU81318.1'],\n",
       "  [452.907947842285, 455.5, 453.907947842285, 469.5, 73, None],\n",
       "  [94.88078236572493,\n",
       "   170.0,\n",
       "   274.88078236572494,\n",
       "   183.0,\n",
       "   28,\n",
       "   'Lontra_canadensis|XP_032721188.1'],\n",
       "  [452.907947842285, 169.5, 453.907947842285, 183.5, 28, None],\n",
       "  [163.14095001552312,\n",
       "   651.0,\n",
       "   324.1409500155231,\n",
       "   664.0,\n",
       "   102,\n",
       "   'Homo_sapiens|NP_001138397.1'],\n",
       "  [452.907947842285, 650.5, 453.907947842285, 664.5, 102, None],\n",
       "  [190.37348649487734,\n",
       "   92.0,\n",
       "   380.37348649487734,\n",
       "   105.0,\n",
       "   17,\n",
       "   'Puma_yagouaroundi|XP_040339814.1'],\n",
       "  [452.907947842285, 91.5, 453.907947842285, 105.5, 17, None],\n",
       "  [97.28314188140328,\n",
       "   157.0,\n",
       "   306.28314188140325,\n",
       "   170.0,\n",
       "   26,\n",
       "   'Nyctereutes_procyonoides|CAD7693676.1'],\n",
       "  [452.907947842285, 156.5, 453.907947842285, 170.5, 26, None],\n",
       "  [164.1027631170444,\n",
       "   612.0,\n",
       "   332.1027631170444,\n",
       "   625.0,\n",
       "   97,\n",
       "   'Pan_troglodytes|XP_009431729.1'],\n",
       "  [452.907947842285, 611.5, 453.907947842285, 625.5, 97, None],\n",
       "  [175.67122011797576,\n",
       "   742.0,\n",
       "   408.67122011797574,\n",
       "   755.0,\n",
       "   115,\n",
       "   'Colobus_angolensis_palliatus|XP_011809449.1'],\n",
       "  [452.907947842285, 741.5, 453.907947842285, 755.5, 115, None],\n",
       "  [186.52747593914933,\n",
       "   352.0,\n",
       "   350.5274759391493,\n",
       "   365.0,\n",
       "   55,\n",
       "   'Carlito_syrichta|XP_008066755.1'],\n",
       "  [452.907947842285, 351.5, 453.907947842285, 365.5, 55, None],\n",
       "  [166.1269791990065,\n",
       "   534.0,\n",
       "   345.12697919900654,\n",
       "   547.0,\n",
       "   86,\n",
       "   'Hylobates_moloch|XP_032030149.1'],\n",
       "  [452.907947842285, 533.5, 453.907947842285, 547.5, 86, None],\n",
       "  [103.74138466314808,\n",
       "   66.0,\n",
       "   239.7413846631481,\n",
       "   79.0,\n",
       "   15,\n",
       "   'Lynx_pardinus|VFV35827.1'],\n",
       "  [452.907947842285, 65.5, 453.907947842285, 79.5, 15, None],\n",
       "  [167.15523129462898,\n",
       "   482.0,\n",
       "   330.15523129462895,\n",
       "   495.0,\n",
       "   76,\n",
       "   'Sapajus_apella|XP_032132720.1'],\n",
       "  [452.907947842285, 481.5, 453.907947842285, 495.5, 76, None],\n",
       "  [163.14095001552312, 664.0, 193.14095001552312, 677.0, 103, 'Query'],\n",
       "  [452.907947842285, 663.5, 453.907947842285, 677.5, 103, None],\n",
       "  [173.23595156783608,\n",
       "   508.0,\n",
       "   321.2359515678361,\n",
       "   521.0,\n",
       "   80,\n",
       "   'Pithecia_pithecia|AFU81315.1'],\n",
       "  [452.907947842285, 507.5, 453.907947842285, 521.5, 80, None],\n",
       "  [99.41788264514126,\n",
       "   1.0,\n",
       "   261.4178826451413,\n",
       "   14.0,\n",
       "   5,\n",
       "   'Panthera_tigris|XP_042812427.1'],\n",
       "  [452.907947842285, 0.5, 453.907947842285, 14.5, 5, None],\n",
       "  [163.49984476870534,\n",
       "   404.0,\n",
       "   342.4998447687053,\n",
       "   417.0,\n",
       "   66,\n",
       "   'Aotus_nancymaae|XP_012300528.1'],\n",
       "  [452.907947842285, 403.5, 453.907947842285, 417.5, 66, None],\n",
       "  [170.0996584911518,\n",
       "   248.0,\n",
       "   290.0996584911518,\n",
       "   261.0,\n",
       "   39,\n",
       "   'Sus_scrofa|BAD11809.1'],\n",
       "  [452.907947842285, 247.5, 453.907947842285, 261.5, 39, None],\n",
       "  [99.90965538652593,\n",
       "   14.0,\n",
       "   252.9096553865259,\n",
       "   27.0,\n",
       "   6,\n",
       "   'Panthera_leo|XP_042811978.1'],\n",
       "  [452.907947842285, 13.5, 453.907947842285, 27.5, 6, None],\n",
       "  [102.23253647935424,\n",
       "   79.0,\n",
       "   275.2325364793542,\n",
       "   92.0,\n",
       "   16,\n",
       "   'Lynx_canadensis|XP_030185366.1'],\n",
       "  [452.907947842285, 78.5, 453.907947842285, 92.5, 16, None],\n",
       "  [162.37224464452035,\n",
       "   560.0,\n",
       "   364.3722446445204,\n",
       "   573.0,\n",
       "   88,\n",
       "   'Nomascus_leucogenys|XP_012357180.1'],\n",
       "  [452.907947842285, 559.5, 453.907947842285, 573.5, 88, None],\n",
       "  [103.59608817137534,\n",
       "   40.0,\n",
       "   314.59608817137536,\n",
       "   53.0,\n",
       "   12,\n",
       "   'Prionailurus_bengalensis|XP_043449495.1'],\n",
       "  [452.907947842285, 39.5, 453.907947842285, 53.5, 12, None],\n",
       "  [149.93511331884505,\n",
       "   365.0,\n",
       "   346.9351133188451,\n",
       "   378.0,\n",
       "   57,\n",
       "   'Propithecus_coquereli|XP_012495571.1'],\n",
       "  [452.907947842285, 364.5, 453.907947842285, 378.5, 57, None],\n",
       "  [166.1381558522198,\n",
       "   547.0,\n",
       "   311.1381558522198,\n",
       "   560.0,\n",
       "   87,\n",
       "   'Hylobates_agilis|AFU81307.1'],\n",
       "  [452.907947842285, 546.5, 453.907947842285, 560.5, 87, None],\n",
       "  [173.2458863706923,\n",
       "   729.0,\n",
       "   324.2458863706923,\n",
       "   742.0,\n",
       "   114,\n",
       "   'Colobus_guereza|AFU81305.1'],\n",
       "  [452.907947842285, 728.5, 453.907947842285, 742.5, 114, None],\n",
       "  [154.0785470350823,\n",
       "   573.0,\n",
       "   308.0785470350823,\n",
       "   586.0,\n",
       "   91,\n",
       "   'Pongo_pygmaeus|AFU81316.1'],\n",
       "  [452.907947842285, 572.5, 453.907947842285, 586.5, 91, None],\n",
       "  [156.806271344303,\n",
       "   521.0,\n",
       "   289.806271344303,\n",
       "   534.0,\n",
       "   81,\n",
       "   'Alouatta_sara|AFU81300.1'],\n",
       "  [452.907947842285, 520.5, 453.907947842285, 534.5, 81, None],\n",
       "  [162.2822104936355,\n",
       "   391.0,\n",
       "   336.2822104936355,\n",
       "   404.0,\n",
       "   64,\n",
       "   'Callithrix_jacchus|XP_008984967.1'],\n",
       "  [452.907947842285, 390.5, 453.907947842285, 404.5, 64, None],\n",
       "  [172.6324122943185,\n",
       "   755.0,\n",
       "   339.6324122943185,\n",
       "   768.0,\n",
       "   117,\n",
       "   'Miopithecus_talapoin|AFU81311.1'],\n",
       "  [452.907947842285, 754.5, 453.907947842285, 768.5, 117, None],\n",
       "  [101.78547035082273,\n",
       "   53.0,\n",
       "   245.78547035082272,\n",
       "   66.0,\n",
       "   13,\n",
       "   'Felis_catus|XP_023094481.1'],\n",
       "  [452.907947842285, 52.5, 453.907947842285, 66.5, 13, None],\n",
       "  [97.1707544240919,\n",
       "   27.0,\n",
       "   271.1707544240919,\n",
       "   40.0,\n",
       "   9,\n",
       "   'Acinonyx_jubatus|XP_026897378.1'],\n",
       "  [452.907947842285, 26.5, 453.907947842285, 40.5, 9, None],\n",
       "  [164.63054951878297,\n",
       "   495.0,\n",
       "   340.63054951878297,\n",
       "   508.0,\n",
       "   78,\n",
       "   'Plecturocebus_moloch|AFU81302.1'],\n",
       "  [452.907947842285, 494.5, 453.907947842285, 508.5, 78, None],\n",
       "  [173.69357342440233,\n",
       "   716.0,\n",
       "   388.6935734244023,\n",
       "   729.0,\n",
       "   112,\n",
       "   'Piliocolobus_tephrosceles|XP_023047606.1'],\n",
       "  [452.907947842285, 715.5, 453.907947842285, 729.5, 112, None],\n",
       "  [158.95156783607575,\n",
       "   599.0,\n",
       "   314.95156783607575,\n",
       "   612.0,\n",
       "   94,\n",
       "   'Gorilla_gorilla|NP_001266690.1'],\n",
       "  [452.907947842285, 598.5, 453.907947842285, 612.5, 94, None],\n",
       "  [181.05836696678048,\n",
       "   690.0,\n",
       "   365.05836696678045,\n",
       "   703.0,\n",
       "   109,\n",
       "   'Rhinopithecus_bieti|XP_017735713.1'],\n",
       "  [452.907947842285, 689.5, 453.907947842285, 703.5, 109, None],\n",
       "  [114.65290282520955,\n",
       "   222.0,\n",
       "   278.65290282520954,\n",
       "   235.0,\n",
       "   35,\n",
       "   'Manis_javanica|XP_036870729.1'],\n",
       "  [452.907947842285, 221.5, 453.907947842285, 235.5, 35, None],\n",
       "  [169.8475628686743,\n",
       "   430.0,\n",
       "   333.8475628686743,\n",
       "   443.0,\n",
       "   69,\n",
       "   'Lagothrix_lagotricha|AFU81308.1'],\n",
       "  [452.907947842285, 429.5, 453.907947842285, 443.5, 69, None],\n",
       "  [173.40794784228498,\n",
       "   235.0,\n",
       "   453.407947842285,\n",
       "   248.0,\n",
       "   38,\n",
       "   'Balaenoptera_acutorostrata_scammoni|XP_007172774.1'],\n",
       "  [452.907947842285, 234.5, 453.907947842285, 248.5, 38, None],\n",
       "  [186.90127289661598,\n",
       "   820.0,\n",
       "   383.901272896616,\n",
       "   833.0,\n",
       "   129,\n",
       "   'Theropithecus_gelada|XP_025236083.1'],\n",
       "  [452.907947842285, 819.5, 453.907947842285, 833.5, 129, None],\n",
       "  [152.46910897236884,\n",
       "   586.0,\n",
       "   279.4691089723689,\n",
       "   599.0,\n",
       "   92,\n",
       "   'Pongo_abelii|PNJ42385.1'],\n",
       "  [452.907947842285, 585.5, 453.907947842285, 599.5, 92, None],\n",
       "  [173.44085687674635,\n",
       "   261.0,\n",
       "   408.44085687674635,\n",
       "   274.0,\n",
       "   42,\n",
       "   'Ceratotherium_simum_simum|XP_004429677.1'],\n",
       "  [452.907947842285, 260.5, 453.907947842285, 274.5, 42, None],\n",
       "  [98.27786401738591,\n",
       "   183.0,\n",
       "   270.2778640173859,\n",
       "   196.0,\n",
       "   30,\n",
       "   'Mustela_erminea|XP_032211404.1'],\n",
       "  [452.907947842285, 182.5, 453.907947842285, 196.5, 30, None],\n",
       "  [116.95529338714685,\n",
       "   209.0,\n",
       "   302.95529338714687,\n",
       "   222.0,\n",
       "   34,\n",
       "   'Manis_pentadactyla|XP_036732799.1'],\n",
       "  [452.907947842285, 208.5, 453.907947842285, 222.5, 34, None],\n",
       "  [192.24930145917415,\n",
       "   326.0,\n",
       "   390.24930145917415,\n",
       "   339.0,\n",
       "   51,\n",
       "   'Orycteropus_afer_afer|XP_007945509.1'],\n",
       "  [452.907947842285, 325.5, 453.907947842285, 339.5, 51, None],\n",
       "  [175.74821484011176,\n",
       "   872.0,\n",
       "   344.74821484011176,\n",
       "   885.0,\n",
       "   135,\n",
       "   'Macaca_mulatta|XP_014988323.2'],\n",
       "  [452.907947842285, 871.5, 453.907947842285, 885.5, 135, None],\n",
       "  [182.06240298044088,\n",
       "   807.0,\n",
       "   388.0624029804409,\n",
       "   820.0,\n",
       "   127,\n",
       "   'Mandrillus_leucophaeus|XP_011839575.1'],\n",
       "  [452.907947842285, 806.5, 453.907947842285, 820.5, 127, None],\n",
       "  [161.2303632412294,\n",
       "   638.0,\n",
       "   320.2303632412294,\n",
       "   651.0,\n",
       "   100,\n",
       "   'synthetic_construct|AAX36236.1'],\n",
       "  [452.907947842285, 637.5, 453.907947842285, 651.5, 100, None],\n",
       "  [169.45700093138777,\n",
       "   768.0,\n",
       "   340.45700093138777,\n",
       "   781.0,\n",
       "   120,\n",
       "   'Chlorocebus_tantalus|AFU81304.1'],\n",
       "  [452.907947842285, 767.5, 453.907947842285, 781.5, 120, None],\n",
       "  [96.48959950325987,\n",
       "   196.0,\n",
       "   293.48959950325985,\n",
       "   209.0,\n",
       "   31,\n",
       "   'Mustela_putorius_furo|XP_004762249.1'],\n",
       "  [452.907947842285, 195.5, 453.907947842285, 209.5, 31, None],\n",
       "  [175.41353616889165,\n",
       "   846.0,\n",
       "   327.41353616889165,\n",
       "   859.0,\n",
       "   131,\n",
       "   'Macaca_sylvanus|AFU81310.1'],\n",
       "  [452.907947842285, 845.5, 453.907947842285, 859.5, 131, None],\n",
       "  [165.5445513815585,\n",
       "   625.0,\n",
       "   323.5445513815585,\n",
       "   638.0,\n",
       "   98,\n",
       "   'Pan_paniscus|NP_001288691.1'],\n",
       "  [452.907947842285, 624.5, 453.907947842285, 638.5, 98, None],\n",
       "  [174.45172306737038,\n",
       "   885.0,\n",
       "   360.4517230673704,\n",
       "   898.0,\n",
       "   136,\n",
       "   'Macaca_nemestrina|XP_011724293.1'],\n",
       "  [452.907947842285, 884.5, 453.907947842285, 898.5, 136, None],\n",
       "  [169.54765600745108,\n",
       "   378.0,\n",
       "   315.5476560074511,\n",
       "   391.0,\n",
       "   62,\n",
       "   'Saguinus_midas|AFU81317.1'],\n",
       "  [452.907947842285, 377.5, 453.907947842285, 391.5, 62, None],\n",
       "  [193.40111766532135,\n",
       "   300.0,\n",
       "   397.40111766532135,\n",
       "   313.0,\n",
       "   48,\n",
       "   'Dasypus_novemcinctus|XP_004466363.1'],\n",
       "  [452.907947842285, 299.5, 453.907947842285, 313.5, 48, None],\n",
       "  [170.77646693573422,\n",
       "   677.0,\n",
       "   357.7764669357342,\n",
       "   690.0,\n",
       "   107,\n",
       "   'Trachypithecus_francoisi|AFU81320.1'],\n",
       "  [452.907947842285, 676.5, 453.907947842285, 690.5, 107, None],\n",
       "  [178.76715305805646,\n",
       "   703.0,\n",
       "   387.76715305805646,\n",
       "   716.0,\n",
       "   110,\n",
       "   'Rhinopithecus_roxellana|XP_030770463.1'],\n",
       "  [452.907947842285, 702.5, 453.907947842285, 716.5, 110, None],\n",
       "  [182.1574045327538,\n",
       "   339.0,\n",
       "   387.1574045327538,\n",
       "   352.0,\n",
       "   54,\n",
       "   'Galeopterus_variegatus|XP_008569442.1'],\n",
       "  [452.907947842285, 338.5, 453.907947842285, 352.5, 54, None]],\n",
       " 'node_areas': {0: [1.0, 1.0, 453.407947842285, 898.0],\n",
       "  1: [3.0, 1.0, 380.3734864948774, 209.0],\n",
       "  2: [32.46103694504812, 1.0, 380.3734864948774, 131.0],\n",
       "  3: [50.72306737038187, 1.0, 380.3734864948774, 105.0],\n",
       "  4: [84.52064576218565, 1.0, 261.4178826451413, 27.0],\n",
       "  5: [96.22198075131946, 1.0, 261.4178826451413, 14.0],\n",
       "  6: [96.22198075131946, 14.0, 252.9096553865259, 27.0],\n",
       "  7: [84.52064576218565, 27.0, 380.3734864948774, 105.0],\n",
       "  8: [88.16361378453895, 27.0, 314.5960881713753, 92.0],\n",
       "  9: [92.22011797578391, 27.0, 271.1707544240919, 40.0],\n",
       "  10: [92.22011797578391, 40.0, 314.5960881713753, 92.0],\n",
       "  11: [94.87954051536789, 40.0, 314.5960881713753, 66.0],\n",
       "  12: [98.53368519093449, 40.0, 314.59608817137536, 53.0],\n",
       "  13: [98.53368519093449, 53.0, 245.78547035082272, 66.0],\n",
       "  14: [94.87954051536789, 66.0, 275.2325364793543, 92.0],\n",
       "  15: [99.55076063334367, 66.0, 239.7413846631481, 79.0],\n",
       "  16: [99.55076063334367, 79.0, 275.2325364793542, 92.0],\n",
       "  17: [88.16361378453895, 92.0, 380.37348649487734, 105.0],\n",
       "  18: [50.72306737038187, 105.0, 285.62278795405155, 131.0],\n",
       "  19: [65.01738590499845, 105.0, 270.1903135672151, 118.0],\n",
       "  20: [65.01738590499845, 118.0, 285.62278795405155, 131.0],\n",
       "  21: [32.46103694504812, 131.0, 309.8140329090345, 209.0],\n",
       "  22: [37.52343992548898, 131.0, 309.81403290903444, 170.0],\n",
       "  23: [75.21049363551691, 131.0, 309.81403290903444, 144.0],\n",
       "  24: [75.21049363551691, 144.0, 306.28314188140325, 170.0],\n",
       "  25: [91.02483700714063, 144.0, 260.97547345544865, 157.0],\n",
       "  26: [91.02483700714063, 157.0, 306.28314188140325, 170.0],\n",
       "  27: [37.52343992548898, 170.0, 293.48959950325985, 209.0],\n",
       "  28: [85.57125116423471, 170.0, 274.88078236572494, 183.0],\n",
       "  29: [85.57125116423471, 183.0, 293.48959950325985, 209.0],\n",
       "  30: [93.22663769015834, 183.0, 270.2778640173859, 196.0],\n",
       "  31: [93.22663769015834, 196.0, 293.48959950325985, 209.0],\n",
       "  32: [3.0, 209.0, 453.407947842285, 898.0],\n",
       "  33: [32.46103694504812, 209.0, 302.95529338714687, 235.0],\n",
       "  34: [98.6820863085998, 209.0, 302.95529338714687, 222.0],\n",
       "  35: [98.6820863085998, 222.0, 278.65290282520954, 235.0],\n",
       "  36: [32.46103694504812, 235.0, 453.407947842285, 898.0],\n",
       "  37: [47.79478422850046, 235.0, 453.407947842285, 261.0],\n",
       "  38: [101.34150884818379, 235.0, 453.407947842285, 248.0],\n",
       "  39: [101.34150884818379, 248.0, 290.0996584911518, 261.0],\n",
       "  40: [47.79478422850046, 261.0, 416.0701645451723, 898.0],\n",
       "  41: [49.95125737348649, 261.0, 408.44085687674635, 287.0],\n",
       "  42: [58.0760633343682, 261.0, 408.44085687674635, 274.0],\n",
       "  43: [58.0760633343682, 274.0, 367.0, 287.0],\n",
       "  44: [49.95125737348649, 287.0, 416.0701645451723, 898.0],\n",
       "  45: [56.42191865880161, 287.0, 397.40111766532135, 339.0],\n",
       "  46: [62.50139708165165, 287.0, 318.19093449239364, 300.0],\n",
       "  47: [62.50139708165165, 300.0, 397.40111766532135, 339.0],\n",
       "  48: [67.98851288419745, 300.0, 397.40111766532135, 313.0],\n",
       "  49: [67.98851288419745, 313.0, 390.24930145917415, 339.0],\n",
       "  50: [96.36541446755665, 313.0, 359.4337162371934, 326.0],\n",
       "  51: [96.36541446755665, 326.0, 390.24930145917415, 339.0],\n",
       "  52: [56.42191865880161, 339.0, 416.07016454517225, 898.0],\n",
       "  53: [61.76373796957466, 339.0, 387.1574045327538, 365.0],\n",
       "  54: [86.56411052468178, 339.0, 387.1574045327538, 352.0],\n",
       "  55: [86.56411052468178, 352.0, 350.5274759391493, 365.0],\n",
       "  56: [61.76373796957466, 365.0, 416.07016454517225, 898.0],\n",
       "  57: [66.24495498292455, 365.0, 346.9351133188451, 378.0],\n",
       "  58: [66.24495498292455, 378.0, 416.07016454517225, 898.0],\n",
       "  59: [106.10027941633032, 378.0, 416.07016454517225, 534.0],\n",
       "  60: [129.42533374728345, 378.0, 416.0701645451723, 495.0],\n",
       "  61: [134.35361688916484, 378.0, 342.4998447687053, 443.0],\n",
       "  62: [138.35423781434335, 378.0, 315.5476560074511, 391.0],\n",
       "  63: [138.35423781434335, 391.0, 342.4998447687053, 443.0],\n",
       "  64: [142.1313256752561, 391.0, 336.2822104936355, 404.0],\n",
       "  65: [142.1313256752561, 404.0, 342.4998447687053, 443.0],\n",
       "  66: [159.10804098106175, 404.0, 342.4998447687053, 417.0],\n",
       "  67: [159.10804098106175, 417.0, 333.8475628686743, 443.0],\n",
       "  68: [161.5998137224464, 417.0, 314.4334057746041, 430.0],\n",
       "  69: [161.5998137224464, 430.0, 333.8475628686743, 443.0],\n",
       "  70: [134.35361688916484, 443.0, 416.07016454517225, 495.0],\n",
       "  71: [143.8084445824278, 443.0, 416.07016454517225, 469.0],\n",
       "  72: [178.34368208630858, 443.0, 416.07016454517225, 456.0],\n",
       "  73: [178.34368208630858, 456.0, 328.5007761564731, 469.0],\n",
       "  74: [143.8084445824278, 469.0, 330.52995963986336, 495.0],\n",
       "  75: [163.67991307047498, 469.0, 330.52995963986336, 482.0],\n",
       "  76: [163.67991307047498, 482.0, 330.15523129462895, 495.0],\n",
       "  77: [129.42533374728345, 495.0, 340.63054951878297, 534.0],\n",
       "  78: [135.18068922694815, 495.0, 340.63054951878297, 508.0],\n",
       "  79: [135.18068922694815, 508.0, 321.2359515678361, 534.0],\n",
       "  80: [141.41664079478423, 508.0, 321.2359515678361, 521.0],\n",
       "  81: [141.41664079478423, 521.0, 289.806271344303, 534.0],\n",
       "  82: [106.10027941633032, 534.0, 408.67122011797574, 898.0],\n",
       "  83: [129.42533374728345, 534.0, 364.3722446445204, 677.0],\n",
       "  84: [138.6230984166408, 534.0, 364.3722446445204, 573.0],\n",
       "  85: [156.5721825520025, 534.0, 345.12697919900654, 560.0],\n",
       "  86: [163.4116733933561, 534.0, 345.12697919900654, 547.0],\n",
       "  87: [163.4116733933561, 547.0, 311.1381558522198, 560.0],\n",
       "  88: [156.5721825520025, 560.0, 364.3722446445204, 573.0],\n",
       "  89: [138.6230984166408, 573.0, 332.1027631170444, 677.0],\n",
       "  90: [143.39490841353617, 573.0, 308.0785470350823, 599.0],\n",
       "  91: [149.83203973921144, 573.0, 308.0785470350823, 586.0],\n",
       "  92: [149.83203973921144, 586.0, 279.4691089723689, 599.0],\n",
       "  93: [143.39490841353617, 599.0, 332.1027631170444, 677.0],\n",
       "  94: [148.88202421608196, 599.0, 314.95156783607575, 612.0],\n",
       "  95: [148.88202421608196, 612.0, 332.1027631170444, 677.0],\n",
       "  96: [152.16733933561005, 612.0, 332.1027631170444, 638.0],\n",
       "  97: [162.1027631170444, 612.0, 332.1027631170444, 625.0],\n",
       "  98: [162.1027631170444, 625.0, 323.5445513815585, 638.0],\n",
       "  99: [152.16733933561005, 638.0, 324.1409500155231, 677.0],\n",
       "  100: [157.74386836386213, 638.0, 320.2303632412294, 651.0],\n",
       "  101: [157.74386836386213, 651.0, 324.1409500155231, 677.0],\n",
       "  102: [161.14095001552312, 651.0, 324.1409500155231, 664.0],\n",
       "  103: [161.14095001552312, 664.0, 193.14095001552312, 677.0],\n",
       "  104: [129.42533374728345, 677.0, 408.6712201179758, 898.0],\n",
       "  105: [153.622167028873, 677.0, 408.67122011797574, 755.0],\n",
       "  106: [158.77398323502015, 677.0, 387.76715305805646, 716.0],\n",
       "  107: [161.90282520956222, 677.0, 357.7764669357342, 690.0],\n",
       "  108: [161.90282520956222, 690.0, 387.76715305805646, 716.0],\n",
       "  109: [171.4358894753182, 690.0, 365.05836696678045, 703.0],\n",
       "  110: [171.4358894753182, 703.0, 387.76715305805646, 716.0],\n",
       "  111: [158.77398323502015, 716.0, 408.67122011797574, 755.0],\n",
       "  112: [161.3551692021111, 716.0, 388.6935734244023, 729.0],\n",
       "  113: [161.3551692021111, 729.0, 408.67122011797574, 755.0],\n",
       "  114: [170.29587084756284, 729.0, 324.2458863706923, 742.0],\n",
       "  115: [170.29587084756284, 742.0, 408.67122011797574, 755.0],\n",
       "  116: [153.622167028873, 755.0, 388.0624029804409, 898.0],\n",
       "  117: [161.37814343371622, 755.0, 339.6324122943185, 768.0],\n",
       "  118: [161.37814343371622, 768.0, 388.0624029804409, 898.0],\n",
       "  119: [163.84756286867432, 768.0, 369.7258615336852, 794.0],\n",
       "  120: [166.42874883576528, 768.0, 340.45700093138777, 781.0],\n",
       "  121: [166.42874883576528, 781.0, 369.7258615336852, 794.0],\n",
       "  122: [163.84756286867432, 794.0, 388.0624029804409, 898.0],\n",
       "  123: [166.8869916175101, 794.0, 388.0624029804409, 859.0],\n",
       "  124: [170.36230984166409, 794.0, 388.0624029804409, 846.0],\n",
       "  125: [174.3517541136293, 794.0, 349.69295249922385, 807.0],\n",
       "  126: [174.3517541136293, 807.0, 388.0624029804409, 846.0],\n",
       "  127: [177.88295560384975, 807.0, 388.0624029804409, 820.0],\n",
       "  128: [177.88295560384975, 820.0, 383.901272896616, 846.0],\n",
       "  129: [180.58708475628688, 820.0, 383.901272896616, 833.0],\n",
       "  130: [180.58708475628688, 833.0, 336.5870847562869, 846.0],\n",
       "  131: [170.36230984166409, 846.0, 327.41353616889165, 859.0],\n",
       "  132: [166.8869916175101, 859.0, 360.4517230673704, 898.0],\n",
       "  133: [169.57994411673394, 859.0, 339.1005898789196, 872.0],\n",
       "  134: [169.57994411673394, 872.0, 360.4517230673704, 898.0],\n",
       "  135: [172.38466314809065, 872.0, 344.74821484011176, 885.0],\n",
       "  136: [172.38466314809065, 885.0, 360.4517230673704, 898.0]},\n",
       " 'nodes': [[88.52483700714063, 155.5, 91.52483700714063, 158.5, 24, None],\n",
       "  [62.51738590499845, 116.5, 65.51738590499845, 119.5, 18, None],\n",
       "  [159.0998137224464, 428.5, 162.0998137224464, 431.5, 67, None],\n",
       "  [82.02064576218565, 42.5625, 85.02064576218565, 45.5625, 3, None],\n",
       "  [138.91664079478423, 519.5, 141.91664079478423, 522.5, 79, None],\n",
       "  [35.02343992548898, 165.25, 38.02343992548898, 168.25, 21, None],\n",
       "  [160.9116733933561, 545.5, 163.9116733933561, 548.5, 85, None],\n",
       "  [172.6005898789196, 864.0, 175.6005898789196, 867.0, 133, None],\n",
       "  [147.31403290903444, 136.0, 150.31403290903444, 139.0, 23, None],\n",
       "  [99.69031356721514, 110.0, 102.69031356721514, 113.0, 19, None],\n",
       "  [140.89490841353617, 602.78125, 143.89490841353617, 605.78125, 89, None],\n",
       "  [178.08708475628688, 831.5, 181.08708475628688, 834.5, 128, None],\n",
       "  [107.12278795405153, 123.0, 110.12278795405153, 126.0, 20, None],\n",
       "  [72.71049363551691, 145.75, 75.71049363551691, 148.75, 22, None],\n",
       "  [180.08708475628688, 838.0, 183.08708475628688, 841.0, 130, None],\n",
       "  [178.57016454517228, 448.0, 181.57016454517228, 451.0, 72, None],\n",
       "  [164.93340577460413, 422.0, 167.93340577460413, 425.0, 68, None],\n",
       "  [175.19295249922385, 799.0, 178.19295249922385, 802.0, 125, None],\n",
       "  [161.17991307047498, 480.5, 164.17991307047498, 483.5, 74, None],\n",
       "  [169.2258615336852, 786.0, 172.2258615336852, 789.0, 121, None],\n",
       "  [166.02995963986336, 474.0, 169.02995963986336, 477.0, 75, None],\n",
       "  [98.84150884818379, 246.5, 101.84150884818379, 249.5, 37, None],\n",
       "  [89.72011797578391, 48.25, 92.72011797578391, 51.25, 8, None],\n",
       "  [167.79587084756284, 740.5, 170.79587084756284, 743.5, 113, None],\n",
       "  [83.07125116423471, 184.75, 86.07125116423471, 187.75, 27, None],\n",
       "  [93.72198075131946, 12.5, 96.72198075131946, 15.5, 4, None],\n",
       "  [93.47547345544862, 149.0, 96.47547345544862, 152.0, 25, None],\n",
       "  [141.69093449239364, 292.0, 144.69093449239364, 295.0, 46, None],\n",
       "  [45.29478422850046,\n",
       "   280.4976501464844,\n",
       "   48.29478422850046,\n",
       "   283.4976501464844,\n",
       "   36,\n",
       "   None],\n",
       "  [90.72663769015834, 194.5, 93.72663769015834, 197.5, 29, None],\n",
       "  [158.64095001552312, 662.5, 161.64095001552312, 665.5, 101, None],\n",
       "  [190.5, 279.0, 193.5, 282.0, 43, None],\n",
       "  [171.93371623719338, 318.0, 174.93371623719338, 321.0, 50, None],\n",
       "  [169.88466314809065, 883.5, 172.88466314809065, 886.5, 134, None],\n",
       "  [180.0007761564731, 461.0, 183.0007761564731, 464.0, 73, None],\n",
       "  [92.38078236572493, 175.0, 95.38078236572493, 178.0, 28, None],\n",
       "  [85.66361378453895, 72.625, 88.66361378453895, 75.625, 7, None],\n",
       "  [155.24386836386213, 652.75, 158.24386836386213, 655.75, 99, None],\n",
       "  [0.5, 186.4447250366211, 3.5, 189.4447250366211, 0, None],\n",
       "  [29.96103694504812, 122.390625, 32.96103694504812, 125.390625, 1, None],\n",
       "  [92.37954051536789, 64.5, 95.37954051536789, 67.5, 10, None],\n",
       "  [160.64095001552312, 656.0, 163.64095001552312, 659.0, 102, None],\n",
       "  [171.8517541136293, 810.375, 174.8517541136293, 813.375, 124, None],\n",
       "  [187.87348649487734, 97.0, 190.87348649487734, 100.0, 17, None],\n",
       "  [94.78314188140328, 162.0, 97.78314188140328, 165.0, 26, None],\n",
       "  [161.6027631170444, 617.0, 164.6027631170444, 620.0, 97, None],\n",
       "  [135.85423781434335, 395.1875, 138.85423781434335, 398.1875, 61, None],\n",
       "  [158.87814343371622,\n",
       "   787.9296875,\n",
       "   161.87814343371622,\n",
       "   790.9296875,\n",
       "   116,\n",
       "   None],\n",
       "  [173.17122011797576, 747.0, 176.17122011797576, 750.0, 115, None],\n",
       "  [29.96103694504812,\n",
       "   250.4988250732422,\n",
       "   32.96103694504812,\n",
       "   253.4988250732422,\n",
       "   32,\n",
       "   None],\n",
       "  [167.86230984166409, 830.6875, 170.86230984166409, 833.6875, 123, None],\n",
       "  [184.02747593914933, 357.0, 187.02747593914933, 360.0, 55, None],\n",
       "  [163.6269791990065, 539.0, 166.6269791990065, 542.0, 86, None],\n",
       "  [101.24138466314808, 71.0, 104.24138466314808, 74.0, 15, None],\n",
       "  [63.744954982924554,\n",
       "   468.71240234375,\n",
       "   66.74495498292455,\n",
       "   471.71240234375,\n",
       "   56,\n",
       "   None],\n",
       "  [154.0721825520025, 555.25, 157.0721825520025, 558.25, 84, None],\n",
       "  [59.26373796957466,\n",
       "   409.606201171875,\n",
       "   62.26373796957466,\n",
       "   412.606201171875,\n",
       "   52,\n",
       "   None],\n",
       "  [164.65523129462898, 487.0, 167.65523129462898, 490.0, 76, None],\n",
       "  [160.64095001552312, 669.0, 163.64095001552312, 672.0, 103, None],\n",
       "  [170.73595156783608, 513.0, 173.73595156783608, 516.0, 80, None],\n",
       "  [96.91788264514126, 6.0, 99.91788264514126, 9.0, 5, None],\n",
       "  [160.99984476870534, 409.0, 163.99984476870534, 412.0, 66, None],\n",
       "  [149.66733933561005, 638.125, 152.66733933561005, 641.125, 95, None],\n",
       "  [167.5996584911518, 253.0, 170.5996584911518, 256.0, 39, None],\n",
       "  [151.122167028873, 749.58984375, 154.122167028873, 752.58984375, 104, None],\n",
       "  [97.40965538652593, 19.0, 100.40965538652593, 22.0, 6, None],\n",
       "  [96.03368519093449, 51.5, 99.03368519093449, 54.5, 11, None],\n",
       "  [159.6027631170444, 623.5, 162.6027631170444, 626.5, 96, None],\n",
       "  [99.73253647935424, 84.0, 102.73253647935424, 87.0, 16, None],\n",
       "  [159.87224464452035, 565.0, 162.87224464452035, 568.0, 88, None],\n",
       "  [101.09608817137534, 45.0, 104.09608817137534, 48.0, 12, None],\n",
       "  [147.43511331884505, 370.0, 150.43511331884505, 373.0, 57, None],\n",
       "  [60.00139708165165, 303.375, 63.00139708165165, 306.375, 45, None],\n",
       "  [163.6381558522198, 552.0, 166.6381558522198, 555.0, 87, None],\n",
       "  [170.7458863706923, 734.0, 173.7458863706923, 737.0, 114, None],\n",
       "  [103.60027941633032,\n",
       "   567.4248046875,\n",
       "   106.60027941633032,\n",
       "   570.4248046875,\n",
       "   58,\n",
       "   None],\n",
       "  [164.3869916175101, 852.21875, 167.3869916175101, 855.21875, 122, None],\n",
       "  [136.1230984166408, 579.015625, 139.1230984166408, 582.015625, 83, None],\n",
       "  [47.45125737348649,\n",
       "   314.49530029296875,\n",
       "   50.45125737348649,\n",
       "   317.49530029296875,\n",
       "   40,\n",
       "   None],\n",
       "  [151.5785470350823, 578.0, 154.5785470350823, 581.0, 91, None],\n",
       "  [156.60804098106175, 418.75, 159.60804098106175, 421.75, 65, None],\n",
       "  [147.33203973921144, 584.5, 150.33203973921144, 587.5, 90, None],\n",
       "  [154.306271344303, 526.0, 157.306271344303, 529.0, 81, None],\n",
       "  [175.38295560384975, 821.75, 178.38295560384975, 824.75, 126, None],\n",
       "  [159.7822104936355, 396.0, 162.7822104936355, 399.0, 64, None],\n",
       "  [170.1324122943185, 760.0, 173.1324122943185, 763.0, 117, None],\n",
       "  [99.28547035082273, 58.0, 102.28547035082273, 61.0, 13, None],\n",
       "  [48.22306737038187, 79.53125, 51.22306737038187, 82.53125, 2, None],\n",
       "  [161.34756286867432, 815.859375, 164.34756286867432, 818.859375, 118, None],\n",
       "  [141.3084445824278, 467.5, 144.3084445824278, 470.5, 70, None],\n",
       "  [94.6707544240919, 32.0, 97.6707544240919, 35.0, 9, None],\n",
       "  [162.13054951878297, 500.0, 165.13054951878297, 503.0, 78, None],\n",
       "  [96.1820863085998, 220.5, 99.1820863085998, 223.5, 33, None],\n",
       "  [159.40282520956222, 691.75, 162.40282520956222, 694.75, 106, None],\n",
       "  [167.07994411673394, 873.75, 170.07994411673394, 876.75, 132, None],\n",
       "  [55.5760633343682, 272.5, 58.5760633343682, 275.5, 41, None],\n",
       "  [156.27398323502015, 711.25, 159.27398323502015, 714.25, 105, None],\n",
       "  [171.19357342440233, 721.0, 174.19357342440233, 724.0, 112, None],\n",
       "  [156.45156783607575, 604.0, 159.45156783607575, 607.0, 94, None],\n",
       "  [178.55836696678048, 695.0, 181.55836696678048, 698.0, 109, None],\n",
       "  [112.15290282520955, 227.0, 115.15290282520955, 230.0, 35, None],\n",
       "  [163.92874883576528, 779.5, 166.92874883576528, 782.5, 119, None],\n",
       "  [167.3475628686743, 435.0, 170.3475628686743, 438.0, 69, None],\n",
       "  [168.9358894753182, 701.5, 171.9358894753182, 704.5, 108, None],\n",
       "  [93.86541446755665, 324.5, 96.86541446755665, 327.5, 49, None],\n",
       "  [170.90794784228498, 240.0, 173.90794784228498, 243.0, 38, None],\n",
       "  [184.40127289661598, 825.0, 187.40127289661598, 828.0, 129, None],\n",
       "  [149.96910897236884, 591.0, 152.96910897236884, 594.0, 92, None],\n",
       "  [65.48851288419745, 314.75, 68.48851288419745, 317.75, 47, None],\n",
       "  [158.8551692021111, 730.75, 161.8551692021111, 733.75, 111, None],\n",
       "  [170.94085687674635, 266.0, 173.94085687674635, 269.0, 42, None],\n",
       "  [95.77786401738591, 188.0, 98.77786401738591, 191.0, 30, None],\n",
       "  [84.06411052468178, 350.5, 87.06411052468178, 353.5, 53, None],\n",
       "  [97.05076063334367, 77.5, 100.05076063334367, 80.5, 14, None],\n",
       "  [114.45529338714685, 214.0, 117.45529338714685, 217.0, 34, None],\n",
       "  [189.74930145917415, 331.0, 192.74930145917415, 334.0, 51, None],\n",
       "  [173.24821484011176, 877.0, 176.24821484011176, 880.0, 135, None],\n",
       "  [131.85361688916484, 431.34375, 134.85361688916484, 434.34375, 60, None],\n",
       "  [179.56240298044088, 812.0, 182.56240298044088, 815.0, 127, None],\n",
       "  [175.84368208630858, 454.5, 178.84368208630858, 457.5, 71, None],\n",
       "  [158.7303632412294, 643.0, 161.7303632412294, 646.0, 100, None],\n",
       "  [139.6313256752561, 407.375, 142.6313256752561, 410.375, 63, None],\n",
       "  [166.95700093138777, 773.0, 169.95700093138777, 776.0, 120, None],\n",
       "  [93.98959950325987, 201.0, 96.98959950325987, 204.0, 31, None],\n",
       "  [126.92533374728345, 470.546875, 129.92533374728345, 473.546875, 59, None],\n",
       "  [172.91353616889165, 851.0, 175.91353616889165, 854.0, 131, None],\n",
       "  [132.68068922694815, 509.75, 135.68068922694815, 512.75, 77, None],\n",
       "  [163.0445513815585, 630.0, 166.0445513815585, 633.0, 98, None],\n",
       "  [53.92191865880161,\n",
       "   356.4906005859375,\n",
       "   56.92191865880161,\n",
       "   359.4906005859375,\n",
       "   44,\n",
       "   None],\n",
       "  [126.92533374728345,\n",
       "   664.302734375,\n",
       "   129.92533374728345,\n",
       "   667.302734375,\n",
       "   82,\n",
       "   None],\n",
       "  [171.95172306737038, 890.0, 174.95172306737038, 893.0, 136, None],\n",
       "  [167.04765600745108, 383.0, 170.04765600745108, 386.0, 62, None],\n",
       "  [190.90111766532135, 305.0, 193.90111766532135, 308.0, 48, None],\n",
       "  [168.27646693573422, 682.0, 171.27646693573422, 685.0, 107, None],\n",
       "  [176.26715305805646, 708.0, 179.26715305805646, 711.0, 110, None],\n",
       "  [179.6574045327538, 344.0, 182.6574045327538, 347.0, 54, None],\n",
       "  [146.38202421608196, 621.0625, 149.38202421608196, 624.0625, 93, None]]}"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Change 'defining string' from protein to protein name/species name/taxid!!!!!!!!!!!!!!\n",
    "#For accession replace accession + name!!!\n",
    "\n",
    "\n",
    "msa_tree.render('msa_tree_midpoint.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#IF you want to add the taxids you need to add the one of the query too\n",
    "#EXCEPTION!! SORRY THE FASTA FORMAT SHOULD PRESENT PROTEIN ID, FORMAT:\n",
    "#>gi|2765634|emb|Z78509.1|PPZ78509 P.pearcei 5.8S rRNA gene and ITS1 and ITS2 DNA\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ete3 import Tree\n",
    "\n",
    "\n",
    "t = Tree( \"((a,b),c);\" )\n",
    "t.render(\"mytree.png\", w=183, units=\"mm\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "newick_string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ete3 import Tree\n",
    "\n",
    "t = Tree(newick_string, quoted_node_names=True, format=1)\n",
    "t\n",
    "#t.render('my_newick_tree.png', w= 200, units='mm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[163843, 109228, 98314, 98318, 109229, 163861, 196630, 1081385, 1704005, 65606, 65607, 65609, 65610, 65613, 65614, 65615, 65616, 65618, 65621, 65624, 65626, 65628, 65629, 65631, 65632, 65634, 1851651, 458857, 458858, 458859, 458860, 458861, 458862, 1048687, 1048688, 655476, 1649342, 2750743, 1851652, 2588810, 92866, 1392492, 65694, 1851653, 190511, 2162857, 2162858, 2162859, 2162860, 2162861, 2162862, 2162864, 2162865, 2162866, 2162901, 2162903, 2162904, 1147103, 1147105, 1147106, 655592, 491753, 557290, 491755, 1048812, 1048813, 1294088, 48018, 327948, 327957, 1966371, 1966372, 2588968, 2588969, 1605932, 1966381, 1966382, 1966383, 1966384, 1966385, 1513475, 131388, 262468, 491852, 491853, 1048917, 1048918, 1048919, 1048920, 1048921, 1048922, 1048923, 1081694, 491872, 491874, 491876, 491878, 1343853, 1343854, 491888]\n"
     ]
    }
   ],
   "source": [
    "from ete3 import NCBITaxa\n",
    "ncbi = NCBITaxa()\n",
    "descendants = ncbi.get_descendant_taxa(40674)\n",
    "# to print names of the taxIDs in the descendants\n",
    "print((descendants[:100]))\n",
    "\n",
    "\n",
    "#Very fast after first time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to check if the Taxid is a species one (so not to be parsed to the lineage)\n",
    "\n",
    "def check_if_species(list_of_taxa):\n",
    "    not_species = []\n",
    "    ranks = ncbi.get_rank(list_of_taxa)\n",
    "    for taxid in list_of_taxa:\n",
    "        if ranks[int(taxid)] != 'species':\n",
    "            not_species.append(taxid)\n",
    "    return not_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1, 131567, 2759, 33154, 33208, 6072, 33213, 33511, 7711, 89593, 7742, 7776, 117570, 117571, 8287, 1338369, 32523, 32524, 40674, 32525, 9347, 1437010, 314146, 9443, 376913, 314293, 9526, 314295, 9604, 207598, 9605, 9606, 1, 131567, 2759, 33154, 33208, 6072, 33213, 33511, 7711, 89593, 7742, 7776, 117570, 117571, 8287, 1338369, 32523, 32524, 40674, 32525, 9347, 1437010, 314146, 9443, 376913, 314293, 9526, 314295, 9604, 207598, 9596, 9597, 1, 131567, 2759, 33154, 33208, 6072, 33213, 33511, 7711, 89593, 7742, 7776, 117570, 117571, 8287, 1338369, 32523, 32524, 40674, 32525, 9347, 1437010, 314146, 9443, 376913, 314293, 9526, 314295, 9604, 207598, 9592, 9593, 1, 131567, 2759, 33154, 33208, 6072, 33213, 33511, 7711, 89593, 7742, 7776, 117570, 117571, 8287, 1338369, 32523, 32524, 40674, 32525, 9347, 1437010, 314146, 9443, 376913, 314293, 9526, 314295, 9604, 607660, 9599, 9600, 1, 131567, 2759, 33154, 33208, 6072, 33213, 33511, 7711, 89593, 7742, 7776, 117570, 117571, 8287, 1338369, 32523, 32524, 40674, 32525, 9347, 1437010, 314146, 9443, 376913, 314293, 9526, 314295, 9604, 607660, 9599, 9601, 1, 131567, 2759, 33154, 33208, 6072, 33213, 33511, 7711, 89593, 7742, 7776, 117570, 117571, 8287, 1338369, 32523, 32524, 40674, 32525, 9347, 1437010, 314146, 9443, 376913, 314293, 9526, 314295, 9577, 325165, 61853, 1, 131567, 2759, 33154, 33208, 6072, 33213, 33511, 7711, 89593, 7742, 7776, 117570, 117571, 8287, 1338369, 32523, 32524, 40674, 32525, 9347, 1437010, 314146, 9443, 376913, 314293, 9526, 314294, 9527, 9528, 9539, 9546, 1, 131567, 2759, 33154, 33208, 6072, 33213, 33511, 7711, 89593, 7742, 7776, 117570, 117571, 8287, 1338369, 32523, 32524, 40674, 32525, 9347, 1437010, 314146, 9443, 376913, 314293, 9526, 314294, 9527, 9528, 9539, 9544, 1, 131567, 2759, 33154, 33208, 6072, 33213, 33511, 7711, 89593, 7742, 7776, 117570, 117571, 8287, 1338369, 32523, 32524, 40674, 32525, 9347, 1437010, 314146, 9443, 376913, 314293, 9526, 314294, 9527, 9528, 9539, 9541, 1, 131567, 2759, 33154, 33208, 6072, 33213, 33511, 7711, 89593, 7742, 7776, 117570, 117571, 8287, 1338369, 32523, 32524, 40674, 32525, 9347, 1437010, 314146, 9443, 376913, 314293, 9526, 314294, 9527, 9569, 54136, 54180]\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'check_if_species' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-86-cff099d19251>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspecies\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 15\u001b[0;31m \u001b[0mchecked_species\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_if_species\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspecies\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'check_if_species' is not defined"
     ]
    }
   ],
   "source": [
    "taxid_list = ['9606', '9597', '9593', '9600', '9601', '61853', '9546', '9544', '9541', '54180']\n",
    "\n",
    "#not_species = check_if_species(taxid_list)\n",
    "#print(not_species)\n",
    "\n",
    "\n",
    "species = []\n",
    "for taxid in taxid_list:\n",
    "    lineage_to_search = ncbi.get_lineage(int(taxid))\n",
    "    for spec in lineage_to_search:\n",
    "        species.append(int(spec))\n",
    "\n",
    "print(species)\n",
    "\n",
    "#checked_species = check_if_species(species)\n",
    "\n",
    "\n",
    "ranks = ncbi.get_rank(species)\n",
    "\n",
    "\n",
    "print(ranks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_list = [163843, 109228, 98314, 98318, 109229, 163861, 196630, 1081385, 1704005, 65606, 65607, 65609, 65610, 65613, 65614, 65615, 65616, 65618, 65621, 65624, 65626, 65628, 65629, 65631, 65632, 65634, 1851651, 458857, 458858, 458859, 458860, 458861, 458862, 1048687, 1048688, 655476, 1649342, 2750743, 1851652, 2588810, 92866, 1392492, 65694, 1851653, 190511, 2162857, 2162858, 2162859, 2162860, 2162861, 2162862, 2162864, 2162865, 2162866, 2162901, 2162903, 2162904, 1147103, 1147105, 1147106, 655592, 491753, 557290, 491755, 1048812, 1048813, 1294088, 48018, 327948, 327957, 1966371, 1966372, 2588968, 2588969, 1605932, 1966381, 1966382, 1966383, 1966384, 1966385, 1513475, 131388, 262468, 491852, 491853, 1048917, 1048918, 1048919, 1048920, 1048921, 1048922, 1048923, 1081694, 491872, 491874, 491876, 491878, 1343853, 1343854, 491888]\n",
    "print(len(new_list))"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "ea58ca30de0aa0442eb55bc2eea49e939cd8fed5b96958dd4f8b808b0201cbdd"
  },
  "kernelspec": {
   "display_name": "Python 3.5.6 64-bit ('ete3_2': conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
